TAO/TAO_IDL/fe/idl.ypp

```diff
@@ -3889,6 +3892,138 @@ enumerator :
         }
         ;
 
+map_type_spec :
+        IDL_MAP
+        '<'
+        map_type
+        ','
+        map_type
+        '>'
+        {
+
+          AST_Map *map = 0;
+          Decl_Annotations_Pair *key_type = $3;
+          Decl_Annotations_Pair *val_type = $5;
+
+          /*
+           * Remove sequence marker from scopes stack.
+           */
+          if (idl_global->scopes ().top () == 0)
+            {
+             idl_global->scopes ().pop ();
+            }
+
```

Comment:
So this isn't currently doing anything as the top scope isn't normally supposed to be null in the middle of parsing. In sequence grammar there is a [separate rule pushing a null scope onto the stack before the base type is parsed](https://github.com/DOCGroup/ACE_TAO/blob/1522d0f3572872020a2eb8167d477c0efa965c50/TAO/TAO_IDL/fe/idl.ypp#L4053-L4056). I didn't know why it was doing this though because these comments are not really useful. Looking around for code that checked the top scope for null I finally found:

https://github.com/DOCGroup/ACE_TAO/blob/1522d0f3572872020a2eb8167d477c0efa965c50/TAO/TAO_IDL/ast/ast_recursive.cpp#L184-L189

So this is referring to Section 7.4.1.4.4.4.4 "Constructed Recursive Types and Forward Declarations" in the IDL spec, which says that:

>The IDL syntax allows the generation of recursive structures and unions via members that have a sequence type. For example, the following:
> ```
> struct Foo;
> typedef sequence<Foo> FooSeq;
> struct Foo {
>   long value;
>   FooSeq chain;
> };
> ```
> The forward declaration for the structure enables the definition of the sequence type `FooSeq`, which is used as the type of the recursive member.

I'm wondering if this should also apply to maps. Taking the spec literally at it's word I'm not sure that's the case because the section doesn't mention maps at all. On the other hand the map section basically says maps are the same as sequences but with the contents are accessed with a key, so it would make sense that a rule that applies to a sequence also applies to maps as long as it makes sense.

This is probably a issue that should be filed with the spec, but @mitza-oci, @jwillemsen do you have any thoughts?

Reply:
I agree with treating maps as different kind of a sequence.  If IDL had parameterized types it would be like we had

```
template<typename K, typename V>
using map = sequence<pair<K, V>>;
```

Reply:
https://issues.omg.org/issues/task-force/IDL43#issue-49295

Reply:
Alright, then @tmayoff can you add the same null scope push that sequence does before the `map_type` rules? I would rather this happen some other way so someone doesn't have to go on a scavenger hunt to figure out why it's doing this, like a specifically named variable set on the scope object, but I guess it's easier just to do what sequence is already doing.